ships/0005-local-source-upload.md

```diff
@@ -0,0 +1,250 @@
+
+When the build steps are being executed _(8)_ in the build-pod, the developer will have the execution
+logs being shown in the command-line _(9)_. The developer can interactively watch the container build
+execution and possible errors that might surface.
+
+At the end, when the build-pod is completed successfully _(5)_, the CLI shows the meta information
+about the container image built, and instructions on how to pull and use the image.
+
+#### Watching the Build-Pod
+
+As mentioned before, Shipwright builds are based on a sequence of containers being executed on a
+single pod, the CLI needs to watch _(5)_ over the state changes, and react as soon as it happens. For
+example, the build-pod can be deleted, fail immediately, fail during execution, and such. Moreover,
+as we can observe on the sequence diagram, it also needs to wait for specific conditions before
+streaming data.
+
+During research and development, this requirement was translated into a [watch][client-go-watch]
+instantiated on the build-pod, and a straightforward mechanism to create an event-loop around pod
+[state changes][event-loop-example].  Each change of state will trigger the call of the informed
+closures, and allows wiring [events and reactions in a practical manner][event-loop-usage].
+
+### Upload Mechanism
+
+#### Redirecting Standard Input
+
+The data upload mechanism is the same employed on <code>[kubectl cp][kubectl-cp]</code>, which is
+redirecting the standard-input (<code>STDIN</code>) from the client to the container, inside the
+desired pod.
+
+Analogous to `kubectl cp`, we will be using `tar` to stream data from the client, while
+<code>[kubectl exec][kubectl-exec]</code> on the pod keeps a <code>tar</code> process waiting for
+data on <code>STDIN</code>.
+
+#### Shipwright Build Controller
+
+The Shipwright Build Controller works based on [Tekton Workspaces][tekton-workspaces], and therefore,
+it already defines the /workspace/source mount point for all containers in the build-pod. Tekton also
+adds it's own helper to ["git clone" the repository][controller-git-clone] in place.
+
+With that said, we can simply stop Shipwright Build-Controller from issuing the "`git clone`", and
+instead, add a [simple helper application][controller-waiter] that's able to hold the build-pod
+execution while the data is transferred.
+
+To accomplish this the Build-Controller needs to be able to detect it should replace the "`git
+clone`" procedure by waiting for the upload helper instead. The changes done during research and
+development are [here][controller-waiter-usage].
+
+##### BuildRun Changes
+
+The proposed API changes are the following:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata: {}
+spec:
+  sources:
+    - name: src
+      type: LocalCopy
+```
+
+The Build Controller will inspect for a type `LocalCopy` on the "`sources"` slice, and therefore
+instead of adding the git clone step, it will instead wait for the user to upload.
```

Comment:
a note perhaps detailing the relationship of buildrun sources with https://github.com/shipwright-io/build/blob/f7ed1768b0dc5e863045b0e758544a08c7acc9b8/pkg/apis/build/v1alpha1/build_types.go#L78

Reply:
Good mention. I'm adding the relationship with previous EP, please consider.

Reply:
ok feel free to resolve this one